File: src/jfan/fan/sys/MimeType.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   14 May 08  Brian Frank  Creation
//

package fan.sys;

import java.util.HashMap;
import java.util.Map.Entry;
import java.util.Iterator;

/**
 * MimeType represents the parsed value of a Content-Type
 * header per RFC 2045 section 5.1.
 */
public final class MimeType
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static MimeType fromStr(String s) { return fromStr(s, true); }
  public static MimeType fromStr(String s, boolean checked)
  {
    try
    {
      // common interned mime types
      switch (s.charAt(0))
      {
        case 'i':
          if (s.equals("image/png"))  return imagePng;
          if (s.equals("image/jpeg")) return imageJpeg;
          if (s.equals("image/gif"))  return imageGif;
          break;
        case 't':
          if (s.equals("text/plain")) return textPlain;
          if (s.equals("text/html"))  return textHtml;
          if (s.equals("text/xml"))   return textXml;
          break;
        case 'x':
          if (s.equals("x-directory/normal")) return dir;
          break;
      }

      int slash = s.indexOf('/');
      String media = s.substring(0, slash);
      String sub = s.substring(slash+1, s.length());
      Map params = emptyParams();

      int semi = sub.indexOf(';');
      if (semi > 0)
      {
        params = doParseParams(sub, semi+1);
        sub = sub.substring(0, semi).trim();
      }

      MimeType r  = new MimeType();
      r.str       = s;
      r.mediaType = FanStr.lower(media);
      r.subType   = FanStr.lower(sub);
      r.params    = params.ro();
      return r;
    }
    catch (ParseErr.Val e)
    {
      if (!checked) return null;
      throw e;
    }
    catch (Exception e)
    {
      if (!checked) return null;
      throw ParseErr.make("MimeType",  s).val;
    }
  }

  public static Map parseParams(String s) { return parseParams(s, true); }
  public static Map parseParams(String s, boolean checked)
  {
    try
    {
      return doParseParams(s, 0);
    }
    catch (ParseErr.Val e)
    {
      if (!checked) return null;
      throw e;
    }
    catch (Exception e)
    {
      if (!checked) return null;
      throw ParseErr.make("MimeType params",  s).val;
    }
  }

  private static Map doParseParams(String s, int offset)
  {
    Map params = new Map(Sys.StrType, Sys.StrType);
    params.caseInsensitive(true);
    boolean inQuotes = false;
    int keyStart = offset;
    int valStart = -1;
    int valEnd   = -1;
    int eq       = -1;
    boolean hasEsc = false;
    for (int i = keyStart; i<s.length(); ++i)
    {
      int c = s.charAt(i);

      if (c == '(' && !inQuotes)
        throw ParseErr.make("MimeType", s, "comments not supported").val;

      if (c == '=' && !inQuotes)
      {
        eq = i++;
        while (FanInt.isSpace(s.charAt(i))) ++i;
        if (s.charAt(i) == '"') { inQuotes = true; ++i; }
        else inQuotes = false;
        valStart = i;
      }

      if (eq < 0) continue;

      if (c == '\\' && inQuotes)
      {
        ++i;
        hasEsc = true;
        continue;
      }

      if (c == '"' && inQuotes)
      {
        valEnd = i-1;
        inQuotes = false;
      }

      if (c == ';' && !inQuotes)
      {
        if (valEnd < 0) valEnd = i-1;
        String key = s.substring(keyStart, eq).trim();
        String val = s.substring(valStart, valEnd+1).trim();
        if (hasEsc) val = unescape(val);
        params.set(key, val);
        keyStart = i+1;
        eq = valStart = valEnd = -1;
        hasEsc = false;
      }
    }

    if (keyStart < s.length())
    {
      if (valEnd < 0) valEnd = s.length()-1;
      String key = s.substring(keyStart, eq).trim();
      String val = s.substring(valStart, valEnd+1).trim();
      if (hasEsc) val = unescape(val);
      params.set(key, val);
    }

    return params;
  }

  private static String unescape(String s)
  {
    StringBuilder buf = new StringBuilder(s.length());
    for (int i=0; i<s.length(); ++i)
    {
      int c = s.charAt(i);
      if (c != '\\') buf.append((char)c);
      else if (s.charAt(i+1) == '\\') { buf.append('\\'); i++; }
    }
    return buf.toString();
  }

//////////////////////////////////////////////////////////////////////////
// Extension
//////////////////////////////////////////////////////////////////////////

  public static MimeType forExt(String s)
  {
    if (s == null) return null;
    synchronized (extLock)
    {
      if (extMap == null) extMap = loadExtMap();
      return (MimeType)extMap.get(FanStr.lower(s));
    }
  }

  static HashMap loadExtMap()
  {
    try
    {
      LocalFile f = new LocalFile(new java.io.File(Sys.HomeDir, "lib" + File.sep + "ext2mime.props"));
      Map props = f.readProps();
      HashMap map = new HashMap((int)props.size() * 3);
      Iterator it = props.pairsIterator();
      while (it.hasNext())
      {
        Entry entry = (Entry)it.next();
        String ext  = (String)entry.getKey();
        String mime = (String)entry.getValue();
        try
        {
          map.put(FanStr.lower(ext), fromStr(mime));
        }
        catch (Exception e)
        {
          System.out.println("WARNING: Invalid entry in lib/ext2mime.props: " + ext + ": " + mime);
        }
      }
      return map;
    }
    catch (Exception e)
    {
      System.out.println("WARNING: Cannot load lib/ext2mime.props");
      System.out.println("  " + e);
    }
    return new HashMap();
  }

  static Object extLock = new Object();
  static HashMap extMap;

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public boolean equals(Object obj)
  {
    if (!(obj instanceof MimeType)) return false;
    MimeType x = (MimeType)obj;
    return mediaType.equals(x.mediaType) &&
           subType.equals(x.subType) &&
           params.equals(x.params);
  }

  public int hashCode()
  {
    return mediaType.hashCode() ^
           subType.hashCode() ^
           params.hashCode();
  }

  public long hash()
  {
    return hashCode();
  }

  public String toStr()
  {
    return str;
  }

  public Type type()
  {
    return Sys.MimeTypeType;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public String mediaType()
  {
    return mediaType;
  }

  public String subType()
  {
    return subType;
  }

  public Map params()
  {
    return params;
  }

  public Charset charset()
  {
    String s = (String)params().get("charset");
    if (s == null) return Charset.utf8;
    return Charset.fromStr(s);
  }

//////////////////////////////////////////////////////////////////////////
// Lazy Load
//////////////////////////////////////////////////////////////////////////

  static Map emptyParams()
  {
    Map q = emptyQuery;
    if (q == null)
    {
      q = new Map(Sys.StrType, Sys.StrType);
      q.caseInsensitive(true);
      q = q.toImmutable();
      emptyQuery = q;
    }
    return q;
  }
  static Map emptyQuery;

//////////////////////////////////////////////////////////////////////////
// Predefined
//////////////////////////////////////////////////////////////////////////

  static MimeType predefined(String media, String sub)
  {
    MimeType t = new MimeType();
    t.mediaType = media;
    t.subType = sub;
    t.params = emptyParams();
    t.str = media + "/" + sub;
    return t;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  static final MimeType imagePng   = predefined("image", "png");
  static final MimeType imageGif   = predefined("image", "gif");
  static final MimeType imageJpeg  = predefined("image", "jpeg");
  static final MimeType textPlain  = predefined("text", "plain");
  static final MimeType textHtml   = predefined("text", "html");
  static final MimeType textXml    = predefined("text", "xml");
  static final MimeType dir        = predefined("x-directory", "normal");

  private String mediaType;
  private String subType;
  private Map params;
  private String str;

}

File: src/nfan/fan/sys/MimeType.cs
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   16 May 08  Andy Frank  Creation
//

using System.Collections;

namespace Fan.Sys
{
  /// <summary>
  /// MimeType represents the parsed value of a Content-Type
  /// header per RFC 2045 section 5.1.
  /// </summary>
  public sealed class MimeType : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static MimeType fromStr(string s) { return fromStr(s, true); }
    public static MimeType fromStr(string s, bool check)
    {
      try
      {
        // common interned mime types
        switch (s[0])
        {
          case 'i':
            if (s == "image/png")  return m_imagePng;
            if (s == "image/jpeg") return m_imageJpeg;
            if (s == "image/gif")  return m_imageGif;
            break;
          case 't':
            if (s == "text/plain") return m_textPlain;
            if (s == "text/html")  return m_textHtml;
            if (s == "text/xml")   return m_textXml;
            break;
          case 'x':
            if (s == "x-directory/normal") return m_dir;
            break;
        }

        int slash = s.IndexOf('/');
        string media = s.Substring(0, slash);
        string sub = s.Substring(slash+1);
        Map pars = emptyParams();

        int semi = sub.IndexOf(';');
        if (semi > 0)
        {
          pars = new Map(Sys.StrType, Sys.StrType);
          pars.caseInsensitive(true);
          bool inQuotes = false;
          int keyStart = semi+1;
          int valStart = -1;
          int valEnd   = -1;
          int eq       = -1;
          for (int i = keyStart; i<sub.Length; ++i)
          {
            int c = sub[i];

            if (c == '(' && !inQuotes)
              throw ParseErr.make("MimeType", s, "comments not supported").val;

            if (c == '=' && !inQuotes)
            {
              eq = i++;
              while (FanInt.isSpace(sub[i])) ++i;
              if (sub[i] == '"') { inQuotes = true; ++i; }
              else inQuotes = false;
              valStart = i;
            }

            if (eq < 0) continue;

            if (c == '"' && inQuotes)
            {
              valEnd = i-1;
              inQuotes = false;
            }

            if (c == ';' && !inQuotes)
            {
              if (valEnd < 0) valEnd = i-1;
              string key = sub.Substring(keyStart, eq-keyStart).Trim();
              string val = sub.Substring(valStart, valEnd+1-valStart).Trim();
              pars.set(key, val);
              keyStart = i+1;
              eq = valStart = valEnd = -1;
            }
          }

          if (keyStart < sub.Length)
          {
            if (valEnd < 0) valEnd = sub.Length-1;
            string key = sub.Substring(keyStart, eq-keyStart).Trim();
            string val = sub.Substring(valStart, valEnd+1-valStart).Trim();
            pars.set(key, val);
          }

          sub = sub.Substring(0, semi).Trim();
        }

        MimeType r    = new MimeType();
        r.m_str       = s;
        r.m_mediaType = FanStr.lower(media);
        r.m_subType   = FanStr.lower(sub);
        r.m_params    = pars.ro();
        return r;
      }
      catch (ParseErr.Val e)
      {
        if (!check) return null;
        throw e;
      }
      catch (System.Exception)
      {
        if (!check) return null;
        throw ParseErr.make("MimeType",  s).val;
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // Extension
  //////////////////////////////////////////////////////////////////////////

    public static MimeType forExt(string s)
    {
      if (s == null) return null;
      lock (m_extLock)
      {
        if (m_extMap == null) m_extMap = loadExtMap();
        return (MimeType)m_extMap[FanStr.lower(s)];
      }
    }

    internal static Hashtable loadExtMap()
    {
      try
      {
        LocalFile f = new LocalFile(new System.IO.FileInfo(Sys.HomeDir + File.m_sep + "lib" + File.m_sep + "ext2mime.props"));
        Map props = f.readProps();
        Hashtable map = new Hashtable((int)props.size() * 3);
        IDictionaryEnumerator en = props.pairsIterator();
        while (en.MoveNext())
        {
          string ext  = (string)en.Key;
          string mime = (string)en.Value;
          try
          {
            map[FanStr.lower(ext)] = fromStr(mime);
          }
          catch (System.Exception)
          {
            System.Console.WriteLine("WARNING: Invalid entry in lib/ext2mime.props: " + ext + ": " + mime);
          }
        }
        return map;
      }
      catch (System.Exception e)
      {
        System.Console.WriteLine("WARNING: Cannot load lib/ext2mime.props");
        System.Console.WriteLine("  " + e);
      }
      return new Hashtable();
    }

    internal static object m_extLock = new object();
    internal static Hashtable m_extMap;

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override bool Equals(object obj)
    {
      if (!(obj is MimeType)) return false;
      MimeType x = (MimeType)obj;
      return
        m_mediaType == x.m_mediaType &&
        m_subType == x.m_subType &&
        m_params.Equals(x.m_params);
    }

    public override int GetHashCode()
    {
      return m_mediaType.GetHashCode() ^
             m_subType.GetHashCode() ^
             m_params.GetHashCode();
    }

    public override long hash()
    {
      return GetHashCode();
    }

    public override string toStr()
    {
      return m_str;
    }

    public override Type type()
    {
      return Sys.MimeTypeType;
    }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public string mediaType()
    {
      return m_mediaType;
    }

    public string subType()
    {
      return m_subType;
    }

    public Map @params()
    {
      return m_params;
    }

    public Charset charset()
    {
      string s = (string)m_params.get("charset");
      if (s == null) return Charset.utf8();
      return Charset.fromStr(s);
    }

  //////////////////////////////////////////////////////////////////////////
  // Lazy Load
  //////////////////////////////////////////////////////////////////////////

    internal static Map emptyParams()
    {
      Map q = m_emptyQuery;
      if (q == null)
      {
        q = new Map(Sys.StrType, Sys.StrType);
        q.caseInsensitive(true);
        q = q.toImmutable();
        m_emptyQuery = q;
      }
      return q;
    }
    static Map m_emptyQuery;

  //////////////////////////////////////////////////////////////////////////
  // Predefined
  //////////////////////////////////////////////////////////////////////////

    static MimeType predefined(string media, string sub)
    {
      MimeType t = new MimeType();
      t.m_mediaType = media;
      t.m_subType = sub;
      t.m_params = emptyParams();
      t.m_str = media + "/" + sub;
      return t;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal static MimeType m_imagePng   = predefined("image", "png");
    internal static MimeType m_imageGif   = predefined("image", "gif");
    internal static MimeType m_imageJpeg  = predefined("image", "jpeg");
    internal static MimeType m_textPlain  = predefined("text", "plain");
    internal static MimeType m_textHtml   = predefined("text", "html");
    internal static MimeType m_textXml    = predefined("text", "xml");
    internal static MimeType m_dir        = predefined("x-directory", "normal");

    private string m_mediaType;
    private string m_subType;
    private Map m_params;
    private string m_str;

  }
}